File: selfdrive/ui/qt/onroad/hud.cc
#include "selfdrive/ui/qt/onroad/hud.h"

#include <cmath>

#include <QPainterPath>

#include "selfdrive/ui/qt/onroad/blind_spot_indicator.h"
#include "selfdrive/ui/qt/util.h"

constexpr int SET_SPEED_NA = 255;

HudRenderer::HudRenderer() {}

void HudRenderer::updateState(const UIState &s) {
  is_metric = s.scene.is_metric;
  status = s.status;

  const SubMaster &sm = *(s.sm);
  if (sm.rcv_frame("carState") < s.scene.started_frame) {
    is_cruise_set = false;
    set_speed = SET_SPEED_NA;
    speed = 0.0;
    return;
  }

  const auto &controls_state = sm["controlsState"].getControlsState();
  const auto &car_state = sm["carState"].getCarState();

  // Handle older routes where vCruiseCluster is not set
  set_speed = car_state.getVCruiseCluster() == 0.0 ? controls_state.getVCruiseDEPRECATED() : car_state.getVCruiseCluster();
  is_cruise_set = set_speed > 0 && set_speed != SET_SPEED_NA;
  is_cruise_available = set_speed != -1;

  if (is_cruise_set && !is_metric) {
    set_speed *= KM_TO_MILE;
  }

  // Handle older routes where vEgoCluster is not set
  v_ego_cluster_seen = v_ego_cluster_seen || car_state.getVEgoCluster() != 0.0;
  float v_ego = v_ego_cluster_seen ? car_state.getVEgoCluster() : car_state.getVEgo();
  speed = std::max<float>(0.0f, v_ego * (is_metric ? MS_TO_KPH : MS_TO_MPH));

  // Speed limit sign: convert m/s to display units (km/h or mph)
  float sl_ms = s.scene.nav_speed_limit_ms;
  nav_speed_limit = (sl_ms > 0.0f) ? std::round(sl_ms * (is_metric ? MS_TO_KPH : MS_TO_MPH)) : 0.0f;

  // EOP: Driver pose status (steering-based monitor — no face fields here)
  show_driver_status = sm.rcv_frame("driverPoseState") >= s.scene.started_frame;
  if (show_driver_status) {
    const auto &driver = sm["driverPoseState"].getDriverPoseState();
    attention_prob = driver.getAttentionProb();
  }
  // Driver overlay data from driverStatus (camera-based: face fields live here)
  if (sm.rcv_frame("driverStatus") >= s.scene.started_frame) {
    const auto &fs = sm["driverStatus"].getDriverStatus();
    driver_detected = fs.getFaceDetected();
    driver_forward = fs.getFaceForward();
    driver_x = fs.getFaceX();
    driver_y = fs.getFaceY();
    driver_yaw = fs.getFaceYaw();
    driver_pitch = fs.getFacePitch();
  }

  // BSD / blinker state
  const auto &cs = car_state;
  left_blinker = cs.getLeftBlinker();
  right_blinker = cs.getRightBlinker();
  // EOP: vehicle-native BSD fused with controlsState BSD (radar + Hailo
  // camera) by the shared getBlindSpotSeverity(), so this cannot disagree
  // with the edge bands or the side-camera border. Collapsed to a bool
  // because the HUD only draws presence, not severity. The shared helper
  // also validity-checks controlsState, which this call site did not.
  const BlindSpotSeverity bs = getBlindSpotSeverity(s);
  left_blindspot = bs.left > 0;
  right_blindspot = bs.right > 0;
  bsd_pulse += 0.08f;
  if (bsd_pulse > 1.0f) bsd_pulse = 0.0f;

  // Turn-by-turn instruction (navd/Valhalla). Map/route stay on the phone;
  // the device only shows the next maneuver as an arrow + distance.
  show_nav_instruction = sm.valid("navInstruction") && sm.rcv_frame("navInstruction") >= s.scene.started_frame;
  if (show_nav_instruction) {
    const auto &nav = sm["navInstruction"].getNavInstruction();
    nav_maneuver_type = nav.getManeuverType().cStr();
    nav_maneuver_modifier = nav.getManeuverModifier().cStr();
    nav_primary_text = nav.getManeuverPrimaryText().cStr();
    nav_maneuver_distance_m = nav.getManeuverDistance();
    show_nav_instruction = !nav_maneuver_type.isEmpty() && nav_maneuver_type != "none";
  }
}

void HudRenderer::draw(QPainter &p, const QRect &surface_rect) {
  p.save();

  // Draw header gradient
  QLinearGradient bg(0, UI_HEADER_HEIGHT - (UI_HEADER_HEIGHT / 2.5), 0, UI_HEADER_HEIGHT);
  bg.setColorAt(0, QColor::fromRgbF(0, 0, 0, 0.45));
  bg.setColorAt(1, QColor::fromRgbF(0, 0, 0, 0));
  p.fillRect(0, 0, surface_rect.width(), UI_HEADER_HEIGHT, bg);


  if (is_cruise_available) {
    drawSetSpeed(p, surface_rect);
  }
  drawCurrentSpeed(p, surface_rect);
  if (nav_speed_limit > 0) {
    drawSpeedLimit(p, surface_rect);
  }

  // EOP: Driver pose status indicator (top-right)
  if (show_driver_status) {
    drawDriverStatus(p, surface_rect);
  }

  // BSD red overlay when changing lane into occupied blind spot
  drawBlindSpotOverlay(p, surface_rect);

  // Turn-by-turn maneuver arrow (top-left) — no map, just the next turn
  if (show_nav_instruction) {
    drawNavInstruction(p, surface_rect);
  }

  p.restore();
}

void HudRenderer::drawSetSpeed(QPainter &p, const QRect &surface_rect) {
  // Draw outer box + border to contain set speed
  const QSize default_size = {110, 130};
  QSize set_speed_size = is_metric ? QSize(125, 130) : default_size;
  QRect set_speed_rect(QPoint(35 + (default_size.width() - set_speed_size.width()) / 2, 25), set_speed_size);

  // Draw set speed box
  p.setPen(QPen(QColor(255, 255, 255, 75), 4));
  p.setBrush(QColor(0, 0, 0, 166));
  p.drawRoundedRect(set_speed_rect, 20, 20);

  // Colors based on status
  QColor max_color = QColor(0xa6, 0xa6, 0xa6, 0xff);
  QColor set_speed_color = QColor(0x72, 0x72, 0x72, 0xff);
  if (is_cruise_set) {
    set_speed_color = QColor(255, 255, 255);
    if (status == STATUS_DISENGAGED) {
      max_color = QColor(255, 255, 255);
    } else if (status == STATUS_OVERRIDE) {
      max_color = QColor(0x91, 0x9b, 0x95, 0xff);
    } else {
      max_color = QColor(0x80, 0xd8, 0xa6, 0xff);
    }
  }

  // Draw "MAX" text
  p.setFont(InterFont(24, QFont::DemiBold));
  p.setPen(max_color);
  p.drawText(set_speed_rect.adjusted(0, 15, 0, 0), Qt::AlignTop | Qt::AlignHCenter, tr("MAX"));

  // Draw set speed
  QString setSpeedStr = is_cruise_set ? QString::number(std::nearbyint(set_speed)) : "–";
  p.setFont(InterFont(32, QFont::Bold));
  p.setPen(set_speed_color);
  p.drawText(set_speed_rect.adjusted(0, 42, 0, 0), Qt::AlignTop | Qt::AlignHCenter, setSpeedStr);
}

void HudRenderer::drawCurrentSpeed(QPainter &p, const QRect &surface_rect) {
  QString speedStr = QString::number(std::nearbyint(speed));

  p.setFont(InterFont(100, QFont::Bold));
  drawText(p, surface_rect.center().x(), 120, speedStr);

  p.setFont(InterFont(38));
  drawText(p, surface_rect.center().x(), 165, is_metric ? tr("km/h") : tr("mph"), 200);
}

void HudRenderer::drawDriverStatus(QPainter &p, const QRect &surface_rect) {
  // Small driver status pill in top-right corner
  QString label = driver_detected ? (driver_forward ? tr("DRIVER") : tr("AWAY")) : tr("NO DRIVER");
  QColor bg_color = driver_detected ? (driver_forward ? QColor(0x00, 0xd8, 0x4a, 0xcc) : QColor(0xff, 0xa5, 0x00, 0xcc))
                                      : QColor(0xff, 0x33, 0x33, 0xcc);

  QFont font = InterFont(16, QFont::DemiBold);
  font.setStyleStrategy(QFont::PreferAntialias);
  p.setFont(font);
  QFontMetrics fm(font);
  int text_w = fm.horizontalAdvance(label);
  int pad_x = 8, pad_y = 4;
  int pill_w = text_w + pad_x * 2;
  int pill_h = fm.height() + pad_y * 2;
  int x = surface_rect.width() - pill_w - 20;
  int y = 20;

  QRect pill(x, y, pill_w, pill_h);
  p.setPen(Qt::NoPen);
  p.setBrush(bg_color);
  p.drawRoundedRect(pill, pill_h / 2, pill_h / 2);

  p.setPen(QColor(0xff, 0xff, 0xff, 0xee));
  p.drawText(pill, Qt::AlignCenter, label);

  // Driver bounding box + gaze arrow overlay
  if (driver_detected) {
    int box_size = 70;
    int fx = int((1.0f - driver_x) * surface_rect.width());
    int fy = int(driver_y * surface_rect.height());
    int alpha = int(180 * (1.0f - attention_prob) + 50);

    // Bounding box with std-modulated alpha
    p.setPen(QPen(QColor(0xff, 0xff, 0xff, alpha), 2));
    p.setBrush(Qt::NoBrush);
    p.drawRoundedRect(fx - box_size / 2, fy - box_size / 2, box_size, box_size, 8, 8);

    // Gaze arrow
    float arrow_len = 25.0f;
    int ax = fx + int(arrow_len * sinf(driver_yaw * 3.14159f / 180.0f));
    int ay = fy - int(arrow_len * sinf(driver_pitch * 3.14159f / 180.0f));
    p.setPen(QPen(QColor(0x00, 0xd8, 0x4a, alpha), 2));
    p.drawLine(fx, fy, ax, ay);
    p.drawEllipse(QPoint(ax, ay), 3, 3);
  }
}

void HudRenderer::drawText(QPainter &p, int x, int y, const QString &text, int alpha) {
  QRect real_rect = p.fontMetrics().boundingRect(text);
  real_rect.moveCenter({x, y - real_rect.height() / 2});

  p.setPen(QColor(0xff, 0xff, 0xff, alpha));
  p.drawText(real_rect.x(), real_rect.bottom(), text);
}

void HudRenderer::drawBlindSpotOverlay(QPainter &p, const QRect &surface_rect) {
  // Show red side-bar warning when blinker is on AND blind spot is occupied
  int w = surface_rect.width();
  int h = surface_rect.height();
  int bar_w = 24;

  // Pulsing alpha for visibility
  int alpha = 120 + int(80 * sinf(bsd_pulse * 3.14159f * 2.0f));

  // Left side: left blinker + left blind spot
  if (left_blinker && left_blindspot) {
    QLinearGradient grad(0, 0, bar_w, 0);
    grad.setColorAt(0.0, QColor(255, 0, 0, alpha));
    grad.setColorAt(1.0, QColor(255, 0, 0, 0));
    p.setPen(Qt::NoPen);
    p.setBrush(grad);
    p.drawRect(0, 0, bar_w, h);

    // "BLIND SPOT" text rotated vertically on left edge
    p.save();
    p.translate(bar_w + 4, h / 2);
    p.rotate(-90);
    p.setFont(InterFont(18, QFont::Bold));
    p.setPen(QColor(255, 50, 50, 220));
    p.drawText(QRect(-100, 0, 200, 24), Qt::AlignCenter, tr("BLIND SPOT"));
    p.restore();
  }

  // Right side: right blinker + right blind spot
  if (right_blinker && right_blindspot) {
    QLinearGradient grad(w, 0, w - bar_w, 0);
    grad.setColorAt(0.0, QColor(255, 0, 0, alpha));
    grad.setColorAt(1.0, QColor(255, 0, 0, 0));
    p.setPen(Qt::NoPen);
    p.setBrush(grad);
    p.drawRect(w - bar_w, 0, bar_w, h);

    // "BLIND SPOT" text rotated vertically on right edge
    p.save();
    p.translate(w - bar_w - 4, h / 2);
    p.rotate(90);
    p.setFont(InterFont(18, QFont::Bold));
    p.setPen(QColor(255, 50, 50, 220));
    p.drawText(QRect(-100, 0, 200, 24), Qt::AlignCenter, tr("BLIND SPOT"));
    p.restore();
  }
}

void HudRenderer::drawSpeedLimit(QPainter &p, const QRect &surface_rect) {
  // MUTCD-style speed limit sign: white circle, red border, black number.
  // Positioned to the right of the set-speed box, vertically centred with it.
  // Works on both 7" (1024×600) and 9.3" (1600×600) displays — coordinates are
  // relative to the left edge of the camera widget, not the full screen width.

  const int radius = 28;
  const int cx = 193;   // right of set-speed box (~x=35+110+20+28)
  const int cy = 90;    // vertical centre of set-speed box (25 + 130/2)

  // White fill
  p.setPen(Qt::NoPen);
  p.setBrush(QColor(255, 255, 255));
  p.drawEllipse(QPoint(cx, cy), radius, radius);

  // Red border
  p.setPen(QPen(QColor(255, 0, 0), 4));
  p.setBrush(Qt::NoBrush);
  p.drawEllipse(QPoint(cx, cy), radius, radius);

  // Speed number in black
  QString limitStr = QString::number(static_cast<int>(nav_speed_limit));
  p.setPen(QColor(0, 0, 0));
  p.setFont(InterFont(52, QFont::Bold));
  QRect text_rect(cx - radius, cy - radius, radius * 2, radius * 2);
  p.drawText(text_rect, Qt::AlignCenter, limitStr);
}

void HudRenderer::drawNavInstruction(QPainter &p, const QRect &surface_rect) {
  // Turn-by-turn arrow + distance, top-left. No map/tiles on-device —
  // full route lives in the NavPilot phone app; this is just the next maneuver.
  const int pill_w = 180;
  const int pill_h = 130;
  const int x = 20;
  const int y = 20;
  QRect pill(x, y, pill_w, pill_h);

  p.setPen(Qt::NoPen);
  p.setBrush(QColor(0, 0, 0, 166));
  p.drawRoundedRect(pill, 20, 20);

  // Arrow rotation angle from Valhalla maneuver modifier (0deg = straight up)
  bool is_uturn = nav_maneuver_type == "uturn";
  bool is_arrive = nav_maneuver_type == "arrive";
  float angle = 0.0f;
  if (is_uturn) {
    angle = 180.0f;
  } else if (nav_maneuver_modifier == "slight right") {
    angle = 30.0f;
  } else if (nav_maneuver_modifier == "right") {
    angle = 90.0f;
  } else if (nav_maneuver_modifier == "sharp right") {
    angle = 135.0f;
  } else if (nav_maneuver_modifier == "slight left") {
    angle = -30.0f;
  } else if (nav_maneuver_modifier == "left") {
    angle = -90.0f;
  } else if (nav_maneuver_modifier == "sharp left") {
    angle = -135.0f;
  }

  const int arrow_cx = x + pill_w / 2;
  const int arrow_cy = y + 45;

  if (is_arrive) {
    // Destination pin instead of a directional arrow
    p.setPen(Qt::NoPen);
    p.setBrush(QColor(0x00, 0xd8, 0x4a));
    p.drawEllipse(QPoint(arrow_cx, arrow_cy), 22, 22);
    p.setPen(QPen(QColor(255, 255, 255), 4));
    p.drawLine(arrow_cx, arrow_cy - 10, arrow_cx, arrow_cy + 10);
    p.drawLine(arrow_cx - 10, arrow_cy, arrow_cx + 10, arrow_cy);
  } else {
    p.save();
    p.translate(arrow_cx, arrow_cy);
    p.rotate(angle);
    QPainterPath arrow;
    arrow.moveTo(0, -30);
    arrow.lineTo(20, 0);
    arrow.lineTo(8, 0);
    arrow.lineTo(8, 26);
    arrow.lineTo(-8, 26);
    arrow.lineTo(-8, 0);
    arrow.lineTo(-20, 0);
    arrow.closeSubpath();
    p.setPen(Qt::NoPen);
    p.setBrush(QColor(255, 255, 255));
    p.drawPath(arrow);
    p.restore();
  }

  // Distance to maneuver
  QString dist_str;
  if (is_metric) {
    dist_str = (nav_maneuver_distance_m >= 1000.0f)
      ? QString::number(nav_maneuver_distance_m / 1000.0f, 'f', 1) + " km"
      : QString::number(int(nav_maneuver_distance_m / 10.0f) * 10) + " m";
  } else {
    float feet = nav_maneuver_distance_m * 3.28084f;
    dist_str = (feet >= 1000.0f)
      ? QString::number(feet / 5280.0f, 'f', 1) + " mi"
      : QString::number(int(feet / 10.0f) * 10) + " ft";
  }
  p.setFont(InterFont(28, QFont::Bold));
  p.setPen(QColor(255, 255, 255));
  p.drawText(QRect(x, y + 75, pill_w, 30), Qt::AlignCenter, dist_str);

  // Street/instruction name (truncated to fit)
  if (!nav_primary_text.isEmpty()) {
    QFont font = InterFont(16);
    p.setFont(font);
    QFontMetrics fm(font);
    QString elided = fm.elidedText(nav_primary_text, Qt::ElideRight, pill_w - 16);
    p.setPen(QColor(255, 255, 255, 200));
    p.drawText(QRect(x, y + 105, pill_w, 22), Qt::AlignCenter, elided);
  }
}

File: selfdrive/ui/eop/components/hud.py
"""The classic onroad HUD.

Port of hud.cc. Same elements in the same places: the MAX/set-speed box top
left, the big current speed centred under a header gradient, a MUTCD speed
limit roundel beside the set-speed box, a driver-monitoring pill top right,
the turn-by-turn maneuver card top left, and red bars down whichever edge the
driver is signalling toward while that blind spot is occupied.

Drawn as one widget rather than a renderer called from the camera widget's
paintEvent, which is how the C++ did it. The reason is EGLFS: on this device
the camera surface is a GL widget, and painting QPainter primitives into a
GL widget's paintEvent is exactly the mix that made Qt terminate (see plan
section 4.3). A separate translucent widget stacked over the camera has no
such constraint.
"""

from __future__ import annotations

import math

from openpilot.selfdrive.ui.eop.components.theme import (
  UI_HEADER_HEIGHT,
  WHITE,
  inter,
)
from openpilot.selfdrive.ui.eop.qt import (
  QBrush,
  QColor,
  QFont,
  QFontMetrics,
  QLinearGradient,
  QPainter,
  QPainterPath,
  QPen,
  QPoint,
  QRect,
  Qt,
  QWidget,
  text_width,
)
from openpilot.selfdrive.ui.eop.state import MS_TO_KPH, MS_TO_MPH, Snapshot, UIStatus

M_TO_FT = 3.28084
FT_PER_MILE = 5280.0

# hud.cc geometry
SET_SPEED_POS = (35, 25)
SET_SPEED_SIZE_IMPERIAL = (110, 130)
SET_SPEED_SIZE_METRIC = (125, 130)
SPEED_LIMIT_CENTRE = (193, 90)
SPEED_LIMIT_RADIUS = 28
NAV_CARD = QRect(20, 20, 180, 130)
BSD_BAR_W = 24

# Frames per BSD pulse. Advanced once per snapshot, i.e. at UI_FREQ_HZ.
BSD_PULSE_FRAMES = 25

# Valhalla maneuver modifiers to arrow rotation, degrees clockwise from up.
TURN_ANGLES = {
  "slight right": 30.0, "right": 90.0, "sharp right": 135.0,
  "slight left": -30.0, "left": -90.0, "sharp left": -135.0,
}


class HudOverlay(QWidget):
  """Translucent HUD layer over the camera."""

  def __init__(self, parent=None):
    super().__init__(parent)
    self.setObjectName("hud")
    self.setAttribute(Qt.WA_TransparentForMouseEvents, True)
    self.setAttribute(Qt.WA_TranslucentBackground, True)
    self._snap = Snapshot()
    self._bsd_frame = 0

  def set_snapshot(self, snap: Snapshot) -> None:
    self._snap = snap
    self._bsd_frame = (self._bsd_frame + 1) % BSD_PULSE_FRAMES
    self.update()

  # ---- painting ---------------------------------------------------------

  def paintEvent(self, event):
    s = self._snap
    p = QPainter(self)
    p.setRenderHint(QPainter.Antialiasing)
    p.setRenderHint(QPainter.TextAntialiasing)

    self._draw_header_gradient(p)
    if s.cruise_available:
      self._draw_set_speed(p, s)
    self._draw_current_speed(p, s)
    if s.speed_limit_ms > 0:
      self._draw_speed_limit(p, s)
    if s.driver.valid:
      self._draw_driver(p, s)
    self._draw_blind_spot_bars(p, s)
    if s.nav.valid:
      self._draw_nav(p, s)

  def _draw_header_gradient(self, p: QPainter) -> None:
    g = QLinearGradient(0, UI_HEADER_HEIGHT - UI_HEADER_HEIGHT / 2.5, 0, UI_HEADER_HEIGHT)
    g.setColorAt(0, QColor(0, 0, 0, int(0.45 * 255)))
    g.setColorAt(1, QColor(0, 0, 0, 0))
    p.fillRect(0, 0, self.width(), UI_HEADER_HEIGHT, QBrush(g))

  def _draw_set_speed(self, p: QPainter, s: Snapshot) -> None:
    default_w = SET_SPEED_SIZE_IMPERIAL[0]
    w, h = SET_SPEED_SIZE_METRIC if s.is_metric else SET_SPEED_SIZE_IMPERIAL
    # Centre the wider metric box on the same axis as the narrower one, so
    # the element does not shift when the unit setting changes.
    rect = QRect(SET_SPEED_POS[0] + (default_w - w) // 2, SET_SPEED_POS[1], w, h)

    p.setPen(QPen(QColor(255, 255, 255, 75), 4))
    p.setBrush(QColor(0, 0, 0, 166))
    p.drawRoundedRect(rect, 20, 20)

    max_color = QColor(0xa6, 0xa6, 0xa6)
    value_color = QColor(0x72, 0x72, 0x72)
    if s.cruise_set:
      value_color = WHITE
      if s.status is UIStatus.DISENGAGED:
        max_color = WHITE
      elif s.status is UIStatus.OVERRIDE:
        max_color = QColor(0x91, 0x9b, 0x95)
      else:
        max_color = QColor(0x80, 0xd8, 0xa6)

    p.setFont(inter(24, QFont.DemiBold))
    p.setPen(max_color)
    p.drawText(rect.adjusted(0, 15, 0, 0), Qt.AlignTop | Qt.AlignHCenter, "MAX")

    p.setFont(inter(32, QFont.Bold))
    p.setPen(value_color)
    text = f"{round(s.set_speed):.0f}" if s.cruise_set else "–"
    p.drawText(rect.adjusted(0, 42, 0, 0), Qt.AlignTop | Qt.AlignHCenter, text)

  def _draw_current_speed(self, p: QPainter, s: Snapshot) -> None:
    cx = self.width() // 2
    p.setFont(inter(100, QFont.Bold))
    self._centred_text(p, cx, 120, f"{round(max(0.0, s.speed_display)):.0f}", 255)
    p.setFont(inter(38))
    self._centred_text(p, cx, 165, s.speed_unit, 200)

  @staticmethod
  def _centred_text(p: QPainter, x: int, y: int, text: str, alpha: int) -> None:
    """hud.cc drawText(): centre on x, sit the baseline at y.

    Uses the glyph bounding rect rather than the font's line box, so the
    digits are optically centred instead of centred on the line height.
    """
    r = p.fontMetrics().boundingRect(text)
    r.moveCenter(QPoint(x, y - r.height() // 2))
    p.setPen(QColor(255, 255, 255, alpha))
    p.drawText(r.x(), r.bottom(), text)

  def _draw_speed_limit(self, p: QPainter, s: Snapshot) -> None:
    """MUTCD roundel: white disc, red ring, black number.

    Sits to the right of the set-speed box. Coordinates are relative to this
    widget's left edge, not the screen's, so the element lands correctly
    whatever the panel width is.
    """
    limit = round(s.speed_limit_ms * (MS_TO_KPH if s.is_metric else MS_TO_MPH))
    cx, cy = SPEED_LIMIT_CENTRE
    centre = QPoint(cx, cy)

    p.setPen(Qt.NoPen)
    p.setBrush(WHITE)
    p.drawEllipse(centre, SPEED_LIMIT_RADIUS, SPEED_LIMIT_RADIUS)
    p.setPen(QPen(QColor(255, 0, 0), 4))
    p.setBrush(Qt.NoBrush)
    p.drawEllipse(centre, SPEED_LIMIT_RADIUS, SPEED_LIMIT_RADIUS)

    p.setPen(QColor(0, 0, 0))
    p.setFont(inter(52, QFont.Bold))
    box = QRect(cx - SPEED_LIMIT_RADIUS, cy - SPEED_LIMIT_RADIUS,
                SPEED_LIMIT_RADIUS * 2, SPEED_LIMIT_RADIUS * 2)
    p.drawText(box, Qt.AlignCenter, f"{limit:.0f}")

  def _draw_driver(self, p: QPainter, s: Snapshot) -> None:
    d = s.driver
    if d.face_detected:
      label = "DRIVER" if d.face_forward else "AWAY"
      bg = QColor(0x00, 0xd8, 0x4a, 0xcc) if d.face_forward else QColor(0xff, 0xa5, 0x00, 0xcc)
    else:
      label, bg = "NO DRIVER", QColor(0xff, 0x33, 0x33, 0xcc)

    font = inter(16, QFont.DemiBold)
    font.setStyleStrategy(QFont.PreferAntialias)
    p.setFont(font)
    fm = QFontMetrics(font)
    pad_x, pad_y = 8, 4
    pill_w = text_width(fm, label) + pad_x * 2
    pill_h = fm.height() + pad_y * 2
    pill = QRect(self.width() - pill_w - 20, 20, pill_w, pill_h)

    p.setPen(Qt.NoPen)
    p.setBrush(bg)
    p.drawRoundedRect(pill, pill_h // 2, pill_h // 2)
    p.setPen(QColor(255, 255, 255, 0xee))
    p.drawText(pill, Qt.AlignCenter, label)

    if d.face_detected:
      self._draw_gaze(p, d)

  def _draw_gaze(self, p: QPainter, d) -> None:
    """Face box and gaze vector, faded by how attentive the driver is: the
    more attention, the fainter the box. It is a warning, not a HUD element,
    so it should recede when everything is fine."""
    box = 70
    # The driver camera is mirrored relative to the screen, so face_x is
    # flipped to put the box on the side the driver's head actually is.
    fx = int((1.0 - d.face_x) * self.width())
    fy = int(d.face_y * self.height())
    alpha = int(180 * (1.0 - d.attention_prob) + 50)

    p.setPen(QPen(QColor(255, 255, 255, alpha), 2))
    p.setBrush(Qt.NoBrush)
    p.drawRoundedRect(fx - box // 2, fy - box // 2, box, box, 8, 8)

    arrow = 25.0
    ax = fx + int(arrow * math.sin(math.radians(d.face_yaw)))
    ay = fy - int(arrow * math.sin(math.radians(d.face_pitch)))
    p.setPen(QPen(QColor(0x00, 0xd8, 0x4a, alpha), 2))
    p.drawLine(fx, fy, ax, ay)
    p.drawEllipse(QPoint(ax, ay), 3, 3)

  def _draw_blind_spot_bars(self, p: QPainter, s: Snapshot) -> None:
    """A pulsing red bar plus a vertical BLIND SPOT label, shown only when the
    driver is signalling *into* an occupied blind spot.

    This is deliberately narrower than BlindSpotBands, which shows any
    occupied blind spot regardless of intent. Here the driver has said they
    intend to move into the lane a car is already in, which is the case worth
    shouting about.
    """
    h = self.height()
    w = self.width()
    phase = self._bsd_frame / BSD_PULSE_FRAMES
    alpha = 120 + int(80 * math.sin(phase * math.pi * 2.0))

    left = s.left_blinker and s.blind_spot.left > 0
    right = s.right_blinker and s.blind_spot.right > 0

    if left:
      self._bsd_bar(p, QRect(0, 0, BSD_BAR_W, h), 0, BSD_BAR_W, alpha)
      self._bsd_label(p, BSD_BAR_W + 4, h // 2, -90)
    if right:
      self._bsd_bar(p, QRect(w - BSD_BAR_W, 0, BSD_BAR_W, h), w, w - BSD_BAR_W, alpha)
      self._bsd_label(p, w - BSD_BAR_W - 4, h // 2, 90)

  @staticmethod
  def _bsd_bar(p: QPainter, band: QRect, x_from: int, x_to: int, alpha: int) -> None:
    g = QLinearGradient(x_from, 0, x_to, 0)
    g.setColorAt(0.0, QColor(255, 0, 0, alpha))
    g.setColorAt(1.0, QColor(255, 0, 0, 0))
    p.setPen(Qt.NoPen)
    p.setBrush(QBrush(g))
    p.drawRect(band)

  @staticmethod
  def _bsd_label(p: QPainter, x: int, y: int, rotation: float) -> None:
    p.save()
    p.translate(x, y)
    p.rotate(rotation)
    p.setFont(inter(18, QFont.Bold))
    p.setPen(QColor(255, 50, 50, 220))
    p.drawText(QRect(-100, 0, 200, 24), Qt.AlignCenter, "BLIND SPOT")
    p.restore()

  def _draw_nav(self, p: QPainter, s: Snapshot) -> None:
    """Next maneuver only -- arrow, distance, street name. There is no map on
    this device by design; the route lives in the NavPilot phone app."""
    card = NAV_CARD
    p.setPen(Qt.NoPen)
    p.setBrush(QColor(0, 0, 0, 166))
    p.drawRoundedRect(card, 20, 20)

    cx = card.x() + card.width() // 2
    cy = card.y() + 45
    if s.nav.maneuver_type == "arrive":
      self._draw_destination_pin(p, cx, cy)
    else:
      angle = 180.0 if s.nav.maneuver_type == "uturn" else TURN_ANGLES.get(s.nav.modifier, 0.0)
      self._draw_turn_arrow(p, cx, cy, angle)

    p.setFont(inter(28, QFont.Bold))
    p.setPen(WHITE)
    p.drawText(QRect(card.x(), card.y() + 75, card.width(), 30),
               Qt.AlignCenter, format_distance(s.nav.distance_m, s.is_metric))

    if s.nav.primary_text:
      font = inter(16)
      p.setFont(font)
      elided = QFontMetrics(font).elidedText(s.nav.primary_text, Qt.ElideRight,
                                             card.width() - 16)
      p.setPen(QColor(255, 255, 255, 200))
      p.drawText(QRect(card.x(), card.y() + 105, card.width(), 22),
                 Qt.AlignCenter, elided)

  @staticmethod
  def _draw_destination_pin(p: QPainter, cx: int, cy: int) -> None:
    p.setPen(Qt.NoPen)
    p.setBrush(QColor(0x00, 0xd8, 0x4a))
    p.drawEllipse(QPoint(cx, cy), 22, 22)
    p.setPen(QPen(WHITE, 4))
    p.drawLine(cx, cy - 10, cx, cy + 10)
    p.drawLine(cx - 10, cy, cx + 10, cy)

  @staticmethod
  def _draw_turn_arrow(p: QPainter, cx: int, cy: int, angle: float) -> None:
    p.save()
    p.translate(cx, cy)
    p.rotate(angle)
    path = QPainterPath()
    path.moveTo(0, -30)
    for x, y in ((20, 0), (8, 0), (8, 26), (-8, 26), (-8, 0), (-20, 0)):
      path.lineTo(x, y)
    path.closeSubpath()
    p.setPen(Qt.NoPen)
    p.setBrush(WHITE)
    p.drawPath(path)
    p.restore()


def format_distance(metres: float, is_metric: bool) -> str:
  """Distance to the next maneuver, rounded the way a driver reads it.

  Rounded to 10 below the unit break and to one decimal above, so the number
  changes at a readable rate rather than counting down every metre.
  """
  if is_metric:
    if metres >= 1000.0:
      return f"{metres / 1000.0:.1f} km"
    return f"{int(metres / 10.0) * 10} m"
  feet = metres * M_TO_FT
  if feet >= 1000.0:
    return f"{feet / FT_PER_MILE:.1f} mi"
  return f"{int(feet / 10.0) * 10} ft"
